mlflow/utils/autologging/__init__.py

```diff
@@ -1023,219 +1048,207 @@ def safe_patch_function(*args, **kwargs):
         while exceptions thrown from other parts of `patch_function` are caught and logged as
         warnings.
         """
-        if _is_testing():
-            preexisting_run_for_testing = mlflow.active_run()
-
-        original = gorilla.get_original_attribute(destination, function_name)
-
-        # If the autologging integration associated with this patch is disabled,
-        # call the original function and return
-        if autologging_is_disabled(autologging_integration):
-            return original(*args, **kwargs)
-
-        # Whether or not to exclude auto-autologged content from content explicitly logged via
-        # `mlflow.start_run()`
-        exclusive = get_autologging_config(autologging_integration, "exclusive", False)
+        # Reroute warnings encountered during the patch function implementation to an MLflow event
+        # logger, and enforce silent mode if applicable (i.e. if the corresponding autologging
+        # integration was called with `silent=True`), hiding MLflow event logging statements and
+        # hiding all warnings in the autologging preamble and postamble (i.e. the code surrounding
+        # the user's original / underlying ML function). Non-MLflow warnings are enabled during the
+        # execution of the original / underlying ML function
+        #
+        # Note that we've opted *not* to apply this context manager as a decorator on
+        # `safe_patch_function` because the context-manager-as-decorator pattern uses
+        # `contextlib.ContextDecorator`, which creates generator expressions that cannot be pickled
+        # during model serialization by ML frameworks such as scikit-learn
+        is_silent_mode = get_autologging_config(autologging_integration, "silent", False)
+        with set_mlflow_events_and_warnings_behavior_globally(
+            # MLflow warnings emitted during autologging training sessions are likely not actionable
+            # and result from the autologging implementation invoking another MLflow API.
+            # Accordingly, we reroute these warnings to the MLflow event logger with level WARNING
+            reroute_warnings=True,
+            disable_event_logs=is_silent_mode,
+            disable_warnings=is_silent_mode,
+        ), set_non_mlflow_warnings_behavior_for_current_thread(
+            # non-MLflow Warnings emitted during the autologging preamble (before the original /
+            # underlying ML function is called) and postamble (after the original / underlying ML
+            # function is called) are likely not actionable and result from the autologging
+            # implementation invoking an API from a dependent library. Accordingly, we reroute these
+            # warnings to the MLflow event logger with level WARNING
+            reroute_warnings=True,
+            disable_warnings=is_silent_mode,
+        ):
```

Comment:
For reference, https://docs.python.org/3/howto/logging.html#when-to-use-logging indicates that `warnings.warn()` should be used for avoidable issues that the client application / caller should address by modifying behaviors, while `logger.warning()` should be used to inform users of unexpected errors.

During an MLflow autologging session, MLflow warnings are not actionable as they are a byproduct of the autologging implementation. non-MLflow warnings emitted during the patch function preamble (i.e. before execution of the original / underlying ML function) and postamble (i.e. after execution of the original / underlying ML function) are also a byproduct of the autologging implementation and are not actionable. Accordingly, we reroute both of these to `logger.warning()`.